zokrates_stdlib/stdlib/ecc/edwardsAdd.code

```diff
@@ -0,0 +1,18 @@
+import "ecc/babyjubjubParams.code" as context
+// Add two points on a twisted Edwards curve
+// Curve parameters are defined with the last argument
+// https://en.wikipedia.org/wiki/Twisted_Edwards_curve#Addition_on_twisted_Edwards_curves
+def main(field[2] pt1, field[2] pt2, field[10] context) -> (field[2]):
+
+    field a = context[0]
+    field d = context[1] 
+
+    field u1 = pt1[0]
+    field v1 = pt1[1]
+    field u2 = pt2[0]
+    field v2 = pt2[1]
+
+    field uOut = (u1*v2 + v1*u2) / (1 + d*u1*u2*v1*v2)
```

Comment:
This can be optimized by quite a bit.
It costs 2 constraints for the return arguments, and the rest 21 for the addition.

It can be optimized so the addition (in zokrates code) so the addition costs 14 constraints by reusing some products:

    field a = context[0]
    field d = context[1]

    field u1 = pt1[0]
    field v1 = pt1[1]
    field u2 = pt2[0]
    field v2 = pt2[1]

    field beta = u1*v2
    field gamma = v1*u2
    field delta =  (u1+v1)*(u2+v2)
    field tau = beta*gamma
    field uOut = (beta+gamma) / (1+d*tau)
    field vOut = (delta - beta - gamma) / (1-d*tau)
    return [uOut, vOut]

In pure constraints code this would have been 6 constraints, so there's still room for improvement (maybe the optimizer touches this?)

Reply:
Thanks for pointing that out. 

For now we want to push most of the optimisations to the optimiser. 
I opened an issue https://github.com/Zokrates/ZoKrates/issues/304 for the re-use of intermediary variables.
I also opened an issue to improve constant propagation https://github.com/Zokrates/ZoKrates/issues/307

On additional note: We could also use Montgomery arithmetics in some cases to reduce the number of constraints for incomplete addition to 3.